PL2/PL2-parser/src/main/java/nl/tudelft/pl2016gr2/parser/controller/FullGfaReader.java

```diff
@@ -3,163 +3,229 @@
+    int orientation = 0;
+    while (index < chars.length) {
+      orientation *= SHIFT_BY_BASE_10;
+      orientation += chars[index++] - '0';
+    }
     node.setAlignment(orientation);
   }
 
   /**
-   * Private method for handling an "L" as first character on the line of a gfa file.
-   * @param sc The scanner which is used to parse the file.
+   * Parse the header part of the GFA file, which contains all of the genomes.
+   *
+   * @param chars the characters of the header.
    */
-  private void handleL(Scanner sc) {
-    int parentId = sc.nextInt();
-    sc.next();
-    int childId = sc.nextInt();
-    Node parent = originalGraph.getNode(parentId);
-    Node child = originalGraph.getNode(childId);
-    parent.addOutlink(child.getId());
-    child.addInlink(parent.getId());
-    sc.nextLine();
+  private void parseHeader(char[] chars) {
+    int index = 0;
+    index = skipTillCharacter(chars, index, ':', 2) + 1;
+    int start = index;
+    while (index < chars.length) {
+      index = skipTillCharacter(chars, index, ';', 1);
+      genomes.add(new String(chars, start, index - start));
+      start = index + 1;
+      index += 2;
+    }
   }
-  
+
   /**
-   * This method extracts the single genomes from the ORI tag.
-   * @param line the full ORI tag
-   * @return an arrayList of the names of the genomes.
+   * Skips all characters till the given character is found.
+   *
+   * @param chars      the array of characters.
+   * @param startIndex the index at which to start in the array.
+   * @param ch         the character to skip to.
+   * @param amount     the amount of times to skip the character.
+   * @return the index of the first occurence of the character after the given start index.
    */
-  private ArrayList<String> extractGenomes(String line) {
-    String[] words = line.split(":");
-    if (words[0].endsWith("ORI")) {
-      String[] gens = words[2].split(";");
-      ArrayList<String> gensAr = new ArrayList<>(Arrays.asList(gens));
-      return gensAr;
+  private int skipTillCharacter(char[] chars, int startIndex, char ch, int amount) {
+    int index = startIndex;
+    for (int i = 0; i < amount; i++) {
+      ++index;
+      while (chars[index] != ch) {
+        ++index;
+      }
     }
-    return null;
+    return index;
   }
-  
+
   /**
-   * This method extracts the alignment integer from the START tag.
-   * @param line the full START tag
-   * @return the alignment integer of the tag
+   * Get a node, or create it if it doesn't exist yet.
+   *
+   * @param id the id of the node.
+   * @return the node.
    */
-  private int extractStart(String line) {
-    String[] words = line.split(":");
-    if (words[0].endsWith("START")) {
-      return Integer.parseInt(words[2]);
+  private Node getNode(int id) {
+    Node node = nodes.get(id);
+    if (node == null) {
+      node = new Node(id, 1, new ArrayList<>(), 0);
+      nodes.put(id, node);
     }
-    return 0;
+    return nodes.get(id);
   }
```

Comment:
Since you are returning a new call to `get(id)`, you don't need to store the original return of this call at line `212`. Simply do
```java
if (nodes.get(id) == null) { ... }
return nodes.get(id);
```

Also you are storing the new node at line `214`, which could be piped into the `put()` operation. However this is a minor problem.

We could (should?) benchmark this to see if the double `get()` does not hinder performance, because it is not strictly `O(1)`, but only amortized. Obviously this method will be invoked **a lot**.